feat(OfflineManager): migrate to new Architecture

tyrauber opened the pull request:
- Create NativeOfflineModule.ts TurboModule spec with CodegenTypes
- Migrate Android from Java to Kotlin, extend NativeOfflineModuleSpec
- Migrate iOS to .mm with getTurboModule bridging method
- Update MLRNPackage.kt to register as TurboModule (isTurboModule=true)
- Add MLRNOfflineModule to package.json codegenConfig modulesProvider
- Use EventEmitter pattern (onProgress/onError) instead of NativeEventEmitter
- Export OfflinePackDownloadState from OfflineManager (removed from MLRNModule)
- Align setPackObserver API across platforms (removed Android-only check)
- Convert tests from JavaScript to TypeScript
- Update mocks for TurboModuleRegistry pattern

Closes #1087

tyrauber committed:
feat: migrate OfflineManager to TurboModule architecture

- Create NativeOfflineModule.ts TurboModule spec with CodegenTypes
- Migrate Android from Java to Kotlin, extend NativeOfflineModuleSpec
- Migrate iOS to .mm with getTurboModule bridging method
- Update MLRNPackage.kt to register as TurboModule (isTurboModule=true)
- Add MLRNOfflineModule to package.json codegenConfig modulesProvider
- Use EventEmitter pattern (onProgress/onError) instead of NativeEventEmitter
- Export OfflinePackDownloadState from OfflineManager (removed from MLRNModule)
- Align setPackObserver API across platforms (removed Android-only check)
- Convert tests from JavaScript to TypeScript
- Update mocks for TurboModuleRegistry pattern

Closes #1087

KiwiKilian commented on src/modules/offline/OfflinePack.ts:
```diff
@@ -40,15 +41,24 @@ export class OfflinePack {
+  async pause(): Promise<void> {
+    if (!this.name) {
+      return;
+    }
```
This shouldn't be necessary with a stricter typing. Overall I'm struggling with the identifier beeing the name...

KiwiKilian replied:
TBH I would prefer only supporting IDs. What's the benefit of supporting both? We would need a proper migration for the metadata. The currently only user defined `metadata`/`context` should move down to a key like `data` and we can add our `id` on the root level to prevent interference. I guess the migration must always be checked once when retrieving the the regions.

tyrauber committed:
fix: address PR review feedback for OfflineManager

- Use EventSubscription type from react-native
- Use WithDefault for minZoom/maxZoom defaults in NativeOfflineModule
- Make fields private, remove _ prefix convention
- Use LngLatBounds type for bounds (causes codegen issue - see PR)
- Make metadata type stricter with OfflinePackMetadata interface
- Add test helper methods (resetForTesting, simulateProgressEvent, etc.)
- Move makeNativeBounds into OfflineCreatePackOptions as _makeBounds
- Remove DEFAULT_MIN/MAX_ZOOM constants from Android (handled by WithDefault)

tyrauber committed:
fix: prefix test helper methods with _ for JSDoc parser

- Rename resetForTesting to _resetForTesting
- Rename simulateProgressEvent to _simulateProgressEvent
- Rename simulateErrorEvent to _simulateErrorEvent
- Rename hasRegisteredListeners to _hasRegisteredListeners
- Methods starting with _ are filtered by JSDoc parser as private

tyrauber committed:
fix: LngLatBounds codegen build failure
Inline the tuple type (readonly [CodegenTypes.Double, ...]) directly in NativeOfflineModule.ts, then cast it back to LngLatBounds at the JavaScript layer in OfflinePack.ts.

tyrauber committed:
fix: remove null union from getPackStatus return type

React Native codegen doesn't support union types with null in return values

tyrauber committed:
fix: use codegen-compatible types and fix Android build

TurboModule spec fixes:
- Remove WithDefault (only valid for Fabric components)
- Use Double[] instead of readonly tuple for bounds

Android native fixes:
- Handle nullable callback parameters in OfflineRegionStatusCallback
- Handle nullable bounds from region definition
- Provide fallback for nullable metadata bytes
- Remove deprecated OfflineCallbackName constants from MLRNModule

tyrauber committed:
fix(ios): fix Objective-C++ compilation errors in MLRNOfflineModule

- Replace @import MapLibre with #import <MapLibre/MapLibre.h> in header
- Fix type name from SpecCreatePackOptions to NativeOfflineCreatePackOptions
- Remove .value_or() calls since codegen returns double not std::optional
- Add inline offline callback constants to MLRNModule.m
- Add src/constants.ts with OfflineCallbackName export
- Update exports test

tyrauber committed:
feat(offline): add backwards-compatible UUID support for offline packs

- Native side generates UUID on pack creation, stores in metadata as 'id'
- Rename lookup methods to use identifier-based matching
- First match by 'id' (UUID), fall back to 'name' for backwards compatibility
- Add 'id' to OfflinePackStatus and OfflinePackMetadata types
- Add 'id' and 'identifier' getters to OfflinePack class
- Update OfflinePack methods to use identifier (prefers id over name)
- Add tests for UUID functionality and backwards compatibility

KiwiKilian commented on src/constants.ts:
This is unused/not needed anymore due to the migration.

tyrauber replied:
Awesome. Removed.

tyrauber committed:
refactor: address PR review feedback for offline manager

- Remove unused OfflineCallbackName constant and constants.ts file
- Change _ prefix methods to private in OfflineCreatePackOptions
- Remove test-only methods from OfflineManager (_resetForTesting, _simulateProgressEvent, _simulateErrorEvent, _hasRegisteredListeners)
- Update tests to use bracket notation for accessing private members

tyrauber committed:
refactor(offline): use ID-only approach for offline pack management

- Remove backwards-compatible name fallback in native modules
- Rename getRegionByIdentifier/getPackFromIdentifier to getRegionById/getPackById
- Store and lookup packs by UUID instead of name
- Add id field to NativeOfflinePackStatus type
- createPack() now returns OfflinePack with generated ID
- Add getPackByName() helper for convenience lookups
- New metadata structure: {id, name, data} where data contains user metadata
- Remove deprecated identifier getter from OfflinePack
- Update all tests for ID-based API

BREAKING CHANGE: Offline packs are now identified by UUID instead of name.
Use pack.id for all operations (delete, invalidate, subscribe, etc.).

tyrauber committed:
refactor(offline): rename 'name' parameter to 'id' in native module interface

- Update TypeScript interface parameter names for consistency
- Rename parameter in Android MLRNOfflineModule.kt
- Rename parameter in iOS MLRNOfflineModule.mm
- Affected methods: deletePack, invalidatePack, pausePackDownload,
  resumePackDownload, getPackStatus, setPackObserver

KiwiKilian commented on android/src/main/java/org/maplibre/reactnative/modules/MLRNModule.java:
```diff
@@ -66,16 +66,10 @@ public Map<String, Object> getConstants() {
         offlinePackDownloadStates.put("Active", MLRNOfflineModule.ACTIVE_REGION_DOWNLOAD_STATE);
         offlinePackDownloadStates.put("Complete", MLRNOfflineModule.COMPLETE_REGION_DOWNLOAD_STATE);
 
```
Remove

KiwiKilian reviewed (changes requested):
We are getting there! Still see some API improvement opportunities as noted.

tyrauber committed:
refactor: migrate OfflinePackDownloadState to string literals

- Remove OfflinePackDownloadState constants from MLRNModule.java
- Remove offline callback constants from MLRNModule.m
- Use LngLatBounds type in test file instead of tuple cast
- Migrate state from Int32 to string literals ('inactive' | 'active' | 'complete')
- Remove Unknown state
- Update native modules (Android/iOS) to return string state values

tyrauber committed:
feat(offline): add metadata migration for legacy pack formats

Add migrateMetadata functions to Android (Kotlin) and iOS (Obj-C++)
that migrate legacy metadata format to the current format:

Legacy format: { name: "...", ...userMetadata }
Current format: { id: "uuid", data: { ...userMetadata } }

The migration runs automatically when reading metadata from existing
packs, ensuring backward compatibility with packs created in earlier
versions while standardizing on the new format.

tyrauber committed:
fix(android): handle nullable String parameters in MLRNOfflineModule

Use Elvis operator to provide default values for nullable String
parameters where non-null String is expected:
- setOfflineRegionObserver(name ?: "", ...)
- makeErrorPayload(name ?: "", ...)
- FeatureCollection.fromJson(featureCollectionJSONStr ?: "{}")

tyrauber committed:
Merge pull request #1 from tyrauber/feat/offline-pack-uuid

feat(offline): add UUID support for offline packs

KiwiKilian committed:
Merge branch 'alpha' into feat/new_arch_offline_manager

KiwiKilian committed:
feat: add persistent migrations

KiwiKilian committed:
Merge branch 'alpha' into feat/new_arch_offline_manager

# Conflicts:
#	android/src/main/java/org/maplibre/reactnative/MLRNPackage.kt
#	android/src/main/java/org/maplibre/reactnative/modules/MLRNModule.java
#	src/__tests__/__mocks__/NativeModules.mock.ts

KiwiKilian committed:
fix: udpate example

KiwiKilian committed:
feat: improve example and tests

KiwiKilian committed:
docs: codegen

tyrauber committed:
fix(ios): guard event emitter calls to prevent crash when callback not set
    
Add checks for _eventEmitterCallback before emitting events in MLRNOfflineModule. The crash occurred because MapLibre SDK fires NSNotifications immediately when offline packs resume, but the JS event emitter callback isn't set until React Native connects the listener. This caused std::__throw_bad_function_call when calling emit methods on an empty std::function.

KiwiKilian committed:
fix: deprecations

KiwiKilian committed:
docs: migration

KiwiKilian committed:
Merge branch 'alpha' into feat/new_arch_offline_manager

# Conflicts:
#	docs/content/docs.json

KiwiKilian committed:
feat: improve

autofix-ci[bot] committed:
style: apply automated fixes

KiwiKilian committed:
feat: cleanup

KiwiKilian committed:
style: format
